runtime/src/iree/hal/utils/caching_allocator.c

```diff
@@ -0,0 +1,658 @@
+// The buffer will be retained in the list.
+static void iree_hal_caching_allocator_pool_push_buffer(
+    iree_hal_caching_allocator_pool_t* pool, iree_hal_buffer_t* buffer) {
+  // Retain the buffer; the caller must release it to complete the ownership
+  // transfer.
+  iree_hal_buffer_retain(buffer);
+
+  IREE_ASSERT_LT(pool->free_count, pool->params.max_free_allocation_count);
+
+  // Add to the end of the list (the most recent).
+  iree_host_size_t i = pool->free_count++;
+  pool->free_buffers[i] = buffer;
+
+  // Track that we're now retaining unused memory.
+  pool->free_allocated_size += buffer->allocation_size;
+  IREE_TRACE_PLOT_VALUE_I64(IREE_HAL_CACHING_ALLOCATOR_ID,
+                            pool->free_allocated_size);
+}
+
+// Takes the buffer in the |pool| free list at index |i| and returns ownership.
+// If the list is large this is bad but in most cases it's just a few dozen
+// elements.
+static iree_hal_buffer_t* iree_hal_caching_allocator_pool_take_buffer_at(
+    iree_hal_caching_allocator_pool_t* pool, iree_host_size_t i) {
+  iree_hal_buffer_t* buffer = pool->free_buffers[i];
+  if (i < pool->free_count - 1) {
+    // Shift the list down to keep it dense and in ascending recency order.
+    memmove(&pool->free_buffers[i], &pool->free_buffers[i + 1],
+            (pool->free_count - i - 1) * sizeof(pool->free_buffers[0]));
+  }
+  --pool->free_count;
+  pool->free_allocated_size -= buffer->allocation_size;
+  IREE_TRACE_PLOT_VALUE_I64(IREE_HAL_CACHING_ALLOCATOR_ID,
+                            pool->free_allocated_size);
+  return buffer;
+}
+
+// Scans the |pool| free list for a buffer matching the given requirements and
+// returns ownership.
+static iree_hal_buffer_t* iree_hal_caching_allocator_pool_find_and_take_buffer(
+    iree_hal_caching_allocator_pool_t* pool,
+    const iree_hal_buffer_params_t* params,
+    iree_device_size_t allocation_size) {
+  // Walk backwards so that we check the most recently released buffers first.
+  for (int i = (int)pool->free_count - 1; i >= 0; --i) {
+    // NOTE: we are not currently checking alignment as we don't really have it.
+    // We assume programs will use consistent alignments for a particular heap
+    // (as the heap has a min alignment).
+    iree_hal_buffer_t* buffer = pool->free_buffers[i];
+    if (iree_all_bits_set(iree_hal_buffer_memory_type(buffer), params->type) &&
+        iree_all_bits_set(iree_hal_buffer_allowed_usage(buffer),
+                          params->usage) &&
+        iree_hal_buffer_allocation_size(buffer) == allocation_size) {
+      return iree_hal_caching_allocator_pool_take_buffer_at(pool, i);
+    }
+  }
+  return NULL;  // nothing found
+}
+
+static void iree_hal_caching_allocator_pool_trim_to_size(
+    iree_hal_caching_allocator_pool_t* pool, iree_device_size_t target_size) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  IREE_TRACE_ZONE_APPEND_VALUE(z0, (int64_t)target_size);
+
+  iree_slim_mutex_lock(&pool->mutex);
```

Comment:
I had to read through this a couple of times to understand the locking behavior. Perhaps a comment on why we release the lock for each backing deallocate?